src/Network/Wai/Log/Internal.hs

```diff
@@ -1,39 +1,53 @@
 {-# LANGUAGE RecordWildCards #-}
 module Network.Wai.Log.Internal where
 
-import Control.Monad (when)
-import Data.Aeson.Types (Value, object, emptyObject)
+import Data.Aeson.Types (Value(..), object)
+import Data.ByteString.Builder (Builder)
 import Data.Text (Text)
 import Data.Time.Clock (UTCTime, diffUTCTime, getCurrentTime)
+import Data.UUID (UUID)
+import Data.UUID.V4 (nextRandom)
 import Log (LogLevel)
-import Network.Wai (Middleware)
+import Network.Wai (Application, responseToStream)
 
-import Network.Wai.Log.Options (Options(..), ResponseTime(..))
+import Network.Wai.Log.Options (Options(..), ResponseTime(..), logSendingResponse)
 
 -- | This type matches the one returned by 'getLoggerIO'
 type LoggerIO = UTCTime -> LogLevel -> Text -> Value -> IO ()
 
--- | Create a logging 'Middleware' given a 'LoggerIO' logging function and 'Options'
-logRequestsWith :: LoggerIO -> Options -> Middleware
-logRequestsWith loggerIO Options{..} app req respond = do
-  logIO "Request received" . object . logRequest $ req
+-- | Create a logging 'Middleware' that takes request UUID
+-- given a 'LoggerIO' logging function and 'Options'
+logRequestsWith :: LoggerIO -> Options -> (UUID -> Application) -> Application
+logRequestsWith loggerIO Options{..} mkApp req respond = do
+  uuid <- nextRandom
+  logIO "Request received" $ logRequest uuid req
   tStart <- getCurrentTime
-  app req $ \resp -> do
+  mkApp uuid req $ \resp -> do
     tEnd <- getCurrentTime
-    when logSendingResponse $
-         logIO_ "Sending response"
+    logIO "Sending response" . logSendingResponse $ uuid
     r <- respond resp
     tFull <- getCurrentTime
     let processing = diffUTCTime tEnd  tStart
         full       = diffUTCTime tFull tStart
         times      = ResponseTime{..}
-    logIO "Request complete" . object $ logResponse req resp times
+
+    _ <- case logBody of
+      Nothing ->
+        logIO "Request complete" $ logResponse uuid req resp Null times
+      Just constructBodyToLog -> do
+        let (status, _headers, bodyToIO) = responseToStream resp
+        bodyToIO $ \streamingBody ->
+              let bodyValue :: Builder -> Value
+                  bodyValue = constructBodyToLog status
+
+                  logWithBuilder :: Builder -> IO ()
+                  logWithBuilder b = logIO "Request complete" $
+                    logResponse uuid req resp (bodyValue b) times
+
+              in streamingBody logWithBuilder (return ())
```

Comment:
I don't think this library should be doing any of the `Response` deconstruction at all. That should be for users of the library to implement as they see fit.

See my other comment on `Options` for more context.
With my suggestions there, this would change to be simply:
```suggestion
    case logResponse of
      LogResponseSimple logSimple ->
        logIO "Request complete" . object $ logSimple req resp times
      LogResponseIO logWithIO -> do
        pairs <- logWithIO req resp times
        logIO "Request complete" . object $ pairs
```